File: python-algorithms/trees/sameTree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isSameTreeRecursive(self, p,q):
        if not p and not q:
            return True
        if not p or not q:
            return False
        if p.val != q.val:
            return False
        return self.isSameTreeRecursive(p.left, q.left) and self.isSameTreeRecursive(p.right, q.right)

    def isSameTree(self, p, q):
        pqueue = [p]
        qqueue = [q]

        while pqueue and qqueue:
            pnode = pqueue.pop()
            qnode = qqueue.pop()
            if pnode.val != qnode.val:
                return False
            if pnode.left and qnode.left:
                pqueue.append(pnode.left)
                qqueue.append(qnode.left)
            elif pnode.left or qnode.left:
                return False
            if pnode.right and qnode.right:
                pqueue.append(pnode.right)
                qqueue.append(qnode.right)
            elif pnode.right or qnode.right:
                return False
        return True


q = TreeNode(1)
q.left = TreeNode(2)
q.right = TreeNode(3)
q.right.right = TreeNode(2)


p = TreeNode(1)
p.left = TreeNode(2)
p.right = TreeNode(3)
p.right.right = TreeNode(2)

solution = Solution()
solution.isSameTree(p,q)

File: cpp_algorithms/trees/sameTree.cpp
//
// Created by mddarr on 10/16/20.
//

#include <iostream>
using namespace std;
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution{
public:
    bool isSameTree(TreeNode* p, TreeNode* q){
        if(p == nullptr && q == nullptr){
            return true;
        }else if(p == nullptr | q == nullptr){
            return false;
        }else if(p->val != q-> val){
            return false;
        }else{
            return isSameTree(p->left, q->left) && isSameTree(p->right, q->right );
        }
    }
};
int main(){
    TreeNode root1{1};
    TreeNode left{2};
    TreeNode right{3};

    root1.left = &left;
    root1.right = &right;



    TreeNode root2{1};
    TreeNode left2{2};
    TreeNode right2{3};

    root2.left = &left2;
    root2.right = &right2;

    Solution solution;
    cout << "The two trees are the same " << solution.isSameTree(&root1, &root2);

}
